src/flag_gems/ops/cummin.py

```diff
@@ -0,0 +1,341 @@
+    out_vals = tl.load(out_ptrs, mask=mask)
+    out_indices = tl.load(out_indices_ptrs, mask=mask)
+
+    if pid > 0:
+        partial_min_ptrs = partial_min + pid - 1
+        last_part_min_via_min = tl.load(partial_min_ptrs)
+        partial_min_indices_ptrs = partial_min_indices + pid - 1
+        last_part_min_index_via_min = tl.load(partial_min_indices_ptrs)
+
+        final_vals = tl.minimum(out_vals, last_part_min_via_min)
+        final_indices = tl.where(
+            out_vals <= last_part_min_via_min, out_indices, last_part_min_index_via_min
+        )
+        tl.store(out_ptrs, final_vals.to(out_vals.dtype), mask=mask)
+        tl.store(out_indices_ptrs, final_indices, mask=mask)
+
+
+@libentry()
+@triton.jit(do_not_specialize=["n_elements"])
+def scan_part_min_kernel(
+    inp,
+    out,
+    out_indices,
+    partial_min,
+    partial_min_indices,
+    n_elements,
+    BLOCK_SIZE: tl.constexpr,
+):
+    pid = tl.program_id(0)
+    offset = pid * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+    mask = offset < n_elements
+
+    inp_ptrs = inp + offset
+    inp_vals = tl.load(inp_ptrs, mask=mask)
+    if (
+        tl.constexpr(inp_vals.dtype.is_int64())
+        or tl.constexpr(inp_vals.dtype.is_uint64())
+    ) or tl.constexpr(inp_vals.dtype.is_fp64()):
+        inp_vals = inp_vals
+    elif tl.constexpr(inp_vals.dtype.is_int()):
+        inp_vals = inp_vals.to(tl.int32)
+    else:
+        inp_vals = inp_vals.to(tl.float32)
+    result, cummin_indices = tl_cummin(inp_vals, offset, axis=0)
+
+    # tl.min do not support min_indices_tie_break_right
+    part_min_via_min, part_min_indices_via_min = tl_min_tie_break_right(
+        inp_vals, axis=0
+    )
+    # part_min_via_min = result[BLOCK_SIZE - 1]
+    # part_min_indices_via_min = indices[BLOCK_SIZE - 1]
+
+    out_ptrs = out + offset
+    tl.store(out_ptrs, result, mask=mask)
+
+    in_indices_ptrs = out_indices + cummin_indices
+    indices_mask = cummin_indices < n_elements
+    true_out_indices = tl.load(in_indices_ptrs, mask=indices_mask)
+    out_indices_ptrs = out_indices + offset
+    tl.store(out_indices_ptrs, true_out_indices, mask=mask)
+
+    partial_min_ptrs = partial_min + pid
+    tl.store(partial_min_ptrs, part_min_via_min)
+
+    true_partial_min_indices = tl.load(
+        out_indices + pid * BLOCK_SIZE + part_min_indices_via_min
+    )
+    partial_min_indices_ptrs = partial_min_indices + pid
+    tl.store(partial_min_indices_ptrs, true_partial_min_indices)
+
+
+def scan_then_fan_col(inp, out, out_indices, n_ele, dtype):
+    # TODO(all): tune on target board
+    BLOCK_SIZE = 1024
+    if n_ele <= 1024 * 4:
+        BLOCK_SIZE = triton.next_power_of_2(n_ele)
+    part_num = math.ceil(n_ele / BLOCK_SIZE)
+    partial_min = torch.empty(part_num, dtype=dtype, device=inp.device)
+    partial_min_indices = torch.empty(part_num, dtype=torch.int64, device=inp.device)
+
+    grid = (part_num,)
+    with torch.cuda.device(inp.device):
```

Comment:
I think we can specialize a kernel for `part_num==1`, where `partial_min` and `partial_min_indices` are not needed.